### ISSUE TITLE
Subsymbolic Layer — Clarion's implicit cognition powered by Axon

### ISSUE THREAD

**Issue by daviaws** [Simbolismo-Digital/xoar#7]
<h2>Context</h2>
<p>Everything in Xoar so far is symbolic. Function clauses, ETS rules, operator proposals, preference atoms — all explicit, all inspectable. This is Soar's world.</p>
<p>Clarion's central insight is that cognition has two levels running in parallel:</p>
<ul>
<li><strong>Explicit (top-down)</strong> — rules you can inspect and explain. Xoar has this: compiled codelets + Procedural ETS.</li>
<li><strong>Implicit (bottom-up)</strong> — statistical intuition you can't articulate but that guides behavior. Xoar has nothing here.</li>
</ul>
<p>Axon is Elixir-native, runs on BEAM, produces Nx tensors. It's a GenServer holding weights — which is just another codelet.</p>
<h2>Design</h2>
<h3>The Subsymbolic Codelet</h3>
<p><code>Xoar.Codelets.Subsymbolic</code> — a sensor-mode codelet (ticks on its own interval) that runs a lightweight neural network trained on the drone's own experience.</p>
<p><strong>Input:</strong> current perception state — position, target, nearby obstacles, battery. Read from Perception ETS, encoded as an Nx tensor.</p>
<p><strong>Output:</strong> a preference signal over possible actions — not a proposal, but a <strong>preference bias</strong> that modulates the DecisionCycle.</p>
<pre><code class="language-elixir"># The subsymbolic codelet doesn't propose operators.
# It writes a preference map to the Perception ETS:

Workspace.put(:perception, WME.new(:subsymbolic, :action_preferences, %{
  move_north:     0.82,
  move_northeast: 0.45,
  move_east:      0.91,   # ← the network "feels" east is good here
  evade:          0.12,
  hover:          0.03
}))
</code></pre>
<p>Other codelets — Navigation, Procedural, even HotLearning — can read this and factor it into their proposals. A Navigation codelet that sees the subsymbolic layer strongly preferring east might propose <code>{1, 0}</code> instead of <code>{1, 1}</code>. Or it might ignore it entirely. The explicit level is always in control — the implicit level merely suggests.</p>
<h3>Training loop</h3>
<p>The network trains online, continuously, from episodic memory:</p>
<pre><code>Episodic ETS (history of actions and outcomes)
    ↓ batch sampled by
Subsymbolic Codelet (every N ticks)
    ↓ feeds into
Axon model (lightweight MLP or RNN)
    ↓ updates weights via
Nx gradient descent
    ↓ writes updated preferences to
Perception ETS
</code></pre>
<p><strong>Training signal:</strong> did the action bring the drone closer to target? Did it lead to an impasse? Did it cause oscillation? The episodic log already has <code>from</code>, <code>to</code>, <code>source</code>, and target position — enough to compute a simple reward.</p>
<pre><code class="language-elixir">defp compute_reward(episode) do
  target = Workspace.get(:perception, :drone, :target)
  dist_before = distance(episode.from, target.value)
  dist_after = distance(episode.to, target.value)

  cond do
    dist_after &lt; dist_before -&gt; 1.0    # got closer
    dist_after == dist_before -&gt; -0.1  # wasted a cycle
    dist_after &gt; dist_before -&gt; -0.5   # went backwards
  end
end
</code></pre>
<h3>Architecture of the network</h3>
<p>Start minimal. A 2-layer MLP:</p>
<pre><code class="language-elixir">model =
  Axon.input("state", shape: {nil, input_size})
  |&gt; Axon.dense(32, activation: :relu)
  |&gt; Axon.dense(num_actions, activation: :sigmoid)
</code></pre>
<p>Input: <code>[pos_x, pos_y, target_x, target_y, obstacle_0_x, obstacle_0_y, ..., battery]</code>
Output: preference score per action, 0.0–1.0.</p>
<p>No need for deep networks. The drone's state space is small. The value is not in the network's power — it's in the architecture's ability to have implicit knowledge running alongside explicit rules.</p>
<h3>Clarion parallel</h3>

Clarion | Xoar
-- | --
Subsymbolic level: neural network | Subsymbolic codelet: Axon model in a GenServer
Implicit knowledge: weights | Codelet state: %{model: ..., params: ..., optimizer_state: ...}
Top-down: explicit rules bias network | Procedural ETS rules modulate training signal
Bottom-up: network activations extracted to rules | HotLearning (#4) observes when subsymbolic preferences consistently lead to good outcomes → crystallizes explicit rules
Dual-process interaction | Both levels write to the same Perception ETS, both feed into the same DecisionCycle

<h2>Implementation notes</h2>
<h3>Axon in a GenServer</h3>
<p>Axon models are just Nx computations. Training state (params, optimizer state) lives in the GenServer state. No GPU required — the state space is tiny. CPU Nx backend is fine.</p>
<pre><code class="language-elixir">defmodule Xoar.Codelets.Subsymbolic do
  use Xoar.Codelet, mode: :sensor, tick_ms: 200

  def init_codelet(_opts) do
    model = build_model()
    {init_fn, _predict_fn} = Axon.build(model)
    template = Nx.template({1, @input_size}, :f32)
    params = init_fn.(template, %{})

    %{
      model: model,
      params: params,
      optimizer_state: Polaris.Optimizers.adam(learning_rate: 0.001) |&gt; Polaris.Optimizers.init(params),
      replay_buffer: []
    }
  end

  def tick(state) do
    # 1. Read current perception, predict preferences
    input = encode_perception()
    {_init_fn, predict_fn} = Axon.build(state.model)
    preferences = predict_fn.(state.params, input)

    # 2. Write preferences to ETS
    Workspace.put(:perception, WME.new(:subsymbolic, :action_preferences, decode_preferences(preferences)))

    # 3. Sample from episodic, train
    batch = sample_episodic(state.replay_buffer)
    new_params = train_step(state, batch)

    %{state | params: new_params}
  end
end
</code></pre>
<h3>Performance budget</h3>
<p>The subsymbolic codelet ticks at its own rate (e.g., 200ms) — slower than perception (40ms), independent of the decision cycle. Inference is one forward pass through a small MLP — microseconds on CPU. Training is one gradient step per tick — also microseconds. It should not impact the cognitive cycle at all.</p>
<h3>Weight persistence</h3>
<p>Axon params are Nx tensors — serializable via <code>Nx.serialize/1</code>. Fits into the persistence mechanism (#5):</p>
<ul>
<li>Weights saved to database alongside Procedural rules</li>
<li>Restored on startup</li>
<li>Transferable between cluster nodes — a drone can inherit another drone's intuition</li>
</ul>
<h2>Dependencies</h2>
<ul>
<li>#3 (Procedural codelet) — subsymbolic preferences read by Procedural rules</li>
<li>#4 (HotLearning) — extracts explicit rules from subsymbolic patterns</li>
<li>#5 (Persistence) — saves/restores weights</li>
<li>Axon + Nx + Polaris as mix dependencies</li>
</ul>
<h2>Acceptance criteria</h2>
<ul>
<li>[ ] <code>Xoar.Codelets.Subsymbolic</code> GenServer using Codelet behavior, sensor mode</li>
<li>[ ] Axon MLP model initialized in codelet state</li>
<li>[ ] Forward pass on each tick, writes action preferences to Perception ETS</li>
<li>[ ] Online training from episodic memory with simple reward signal</li>
<li>[ ] At least one existing codelet (Navigation) reads subsymbolic preferences and factors them into proposals</li>
<li>[ ] HotLearning can detect consistent subsymbolic preferences and extract them as Procedural rules</li>
<li>[ ] Weights serializable and restorable via persistence (#5)</li>
<li>[ ] Demo: run the obstacle scenario, observe the subsymbolic layer gradually learning to prefer lateral movement, watch HotLearning eventually crystallize this into an explicit rule</li>
</ul></body></html>
